test/integration/coscheduling_test.go

```diff
@@ -0,0 +1,275 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package integration
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/scheduler"
+	schedapi "k8s.io/kubernetes/pkg/scheduler/apis/config"
+	framework "k8s.io/kubernetes/pkg/scheduler/framework/v1alpha1"
+	st "k8s.io/kubernetes/pkg/scheduler/testing"
+	testutils "k8s.io/kubernetes/test/integration/util"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	"sigs.k8s.io/scheduler-plugins/pkg/coscheduling"
+	"sigs.k8s.io/scheduler-plugins/test/util"
+)
+
+var lowPriority, midPriority, highPriority = int32(0), int32(100), int32(1000)
+
+func TestCoschedulingPlugin(t *testing.T) {
+	registry := framework.Registry{coscheduling.Name: coscheduling.New}
+	profile := schedapi.KubeSchedulerProfile{
+		SchedulerName: v1.DefaultSchedulerName,
+		Plugins: &schedapi.Plugins{
+			QueueSort: &schedapi.PluginSet{
+				Enabled: []schedapi.Plugin{
+					{Name: coscheduling.Name},
+				},
+				Disabled: []schedapi.Plugin{
+					{Name: "*"},
+				},
+			},
+			PreFilter: &schedapi.PluginSet{
+				Enabled: []schedapi.Plugin{
+					{Name: coscheduling.Name},
+				},
+			},
+			Permit: &schedapi.PluginSet{
+				Enabled: []schedapi.Plugin{
+					{Name: coscheduling.Name},
+				},
+			},
+			Unreserve: &schedapi.PluginSet{
+				Enabled: []schedapi.Plugin{
+					{Name: coscheduling.Name},
+				},
+			},
+		},
+	}
+
+	testCtx := util.InitTestSchedulerWithOptions(
+		t,
+		testutils.InitTestMaster(t, "sched-coscheduling", nil),
+		false,
```

Comment:
I'd suggest to set this to `true`, then you don't need L80~L88. Otherwise, this cannot simulate a real scheduler running scenario.

BTW: If I changed to true. The overall time spent for all integration tests are spent as below:

```
    --- PASS: TestCoschedulingPlugin/equal_priority,_sequentially_pg1_meet_min_and_pg2_not_meet_min (90.04s)
    --- PASS: TestCoschedulingPlugin/equal_priority,_not_sequentially_pg1_meet_min_and_pg2_not_meet_min (90.15s)
    --- PASS: TestCoschedulingPlugin/equal_priority,_sequentially_pg1_not_meet_min_and_regular_pods (3.12s)
    --- PASS: TestCoschedulingPlugin/equal_priority,_not_sequentially_pg1_not_meet_min_and_3_regular_pods (3.09s)
    --- PASS: TestCoschedulingPlugin/different_priority,_sequentially_pg1_meet_min_and_pg2_meet_min (84.11s)
    --- PASS: TestCoschedulingPlugin/different_priority,_not_sequentially_pg1_meet_min_and_pg2_meet_min (90.11s)
    --- PASS: TestCoschedulingPlugin/different_priority,_not_sequentially_pg1_meet_min_and_3_regular_pods (3.10s)
    --- PASS: TestCoschedulingPlugin/equal_priority,_not_sequentially_pg1_meet_min_and_p2_p3_not_meet_min (87.13s)
```

Some tests used ~90 seconds, which sounds a bit too much - I assume the first flushoff would take 30 seconds (by default), and each individual test should take less than 40 seconds? Could you help me understand the time cost here? Or, maybe it's wasting some unnecessary scheduling cycles?

Reply:
I find that the flushoff will run every 30s. But it just moves the pods who stays in UnschedulableQ more than 60s. For example: if the scheduler starts in 0s, and the podA is created in 1s and is waiting in `permit` phase. And podA will be timeout in 11s after 10s, Because in the 2th cycle, PodA just stays 49s in UnschedulableQ. So the flush will move PodA in the 3th cycle in 90s. So I think it's the reason that why Some tests used ~90 seconds @Huang-Wei  I think we can improve it by the `Efficient Requeuing of Unschedulable Pods` in the future /cc @alculquicondor 
```go
// flushUnschedulableQLeftover moves pod which stays in unschedulableQ longer than the unschedulableQTimeInterval
// to activeQ.
func (p *PriorityQueue) flushUnschedulableQLeftover() {
	p.lock.Lock()
	defer p.lock.Unlock()

	var podsToMove []*framework.QueuedPodInfo
	currentTime := p.clock.Now()
	for _, pInfo := range p.unschedulableQ.podInfoMap {
		lastScheduleTime := pInfo.Timestamp
		if currentTime.Sub(lastScheduleTime) > unschedulableQTimeInterval {
			podsToMove = append(podsToMove, pInfo)
		}
	}

	if len(podsToMove) > 0 {
		p.movePodsToActiveOrBackoffQueue(podsToMove, UnschedulableTimeout)
	}
}
```

Reply:
SGTM.